test/contrib/flag/test_flag.py

```diff
@@ -107,3 +108,58 @@ def test_zero_grad(model: torch.nn.Module, optimizer: torch.nn.Module,
         for p in group['params']:
             assert torch.count_nonzero(
                 p.grad) == 0, "After running FLAG, the gradients should be 0!"
+
+
+class FLAGLossHistoryCallback(FLAGCallback):
+  def __init__(self):
+    self.loss_history = []
+
+  def on_ascent_step_end(self, step_index, loss, perturb):
+    self.loss_history.append(loss)
+
+
+class FLAGCounterCallback(FLAGCallback):
+  def __init__(self):
+    self.counts = defaultdict(int)
+
+  def on_ascent_step_begin(self, step_index, loss):
+    self.counts['on_ascent_step_begin'] += 1
+
+  def on_ascent_step_end(self, step_index, loss, perturb_data):
+    self.counts['on_ascent_step_end'] += 1
+
+  def on_optimizer_step_begin(self, loss):
```

Comment:
Might it be worthwhile to pass in the optimizer itself here? Same question for `on_optimizer_step_end`.

Reply:
on second thought, I think the `optimizer` is an object that will have the same reference throughout the lifecycle — so I'm not sure whether it actually will change on each step. It might be better to refer to `flag.optimizer` in the callback.